src/names/applications.cpp

```diff
@@ -0,0 +1,159 @@
+            return "";
+    }
+
+    /* NameNamespace values are only ever created internally in the binary
+    and not received externally (except as string).  Thus it should never
+    happen (and if it does, is a severe bug) that we see an unexpected
+    value.  */
+    assert (false);
+}
+
+NameNamespace
+PurportedNamespaceFromName (const std::string& name)
+{
+    const auto slashPos = name.find("/");
+    if (slashPos == std::string::npos)
+    {
+        return NameNamespace::NonStandard;
+    }
+
+    const std::string namespaceStr = name.substr(0, slashPos+1);
+    return NamespaceFromString(namespaceStr);
+}
+
+} // anonymous namespace
+
+NameNamespace
+NamespaceFromName (const std::string& name)
+{
+    const NameNamespace purported = PurportedNamespaceFromName(name);
+    const std::string purportedStr = NamespaceToString(purported);
+    const auto purportedLen = purportedStr.length();
+
+    const std::string label = name.substr(purportedLen);
+
+    if (label.length() < 1)
+    {
+        return NameNamespace::NonStandard;
+    }
+
+    switch (purported)
+    {
+        case NameNamespace::Domain:
+        {
+            // Source: https://github.com/namecoin/proposals/blob/master/ifa-0001.md#keys
+            if (label.length() > 63)
+            {
+                return NameNamespace::NonStandard;
+            }
+
+            // Source: https://github.com/namecoin/proposals/blob/master/ifa-0001.md#keys
+            std::regex domainPattern("(xn--)?[a-z0-9]+(-[a-z0-9]+)*");
+            if (!std::regex_match (label, domainPattern))
+            {
+                return NameNamespace::NonStandard;
+            }
+
+            // Reject digits-only labels
+            // Source: https://github.com/namecoin/proposals/blob/master/ifa-0001.md#keys
+            std::regex digitsOnly("[0-9]+");
+            if (std::regex_match (label, digitsOnly))
+            {
+                return NameNamespace::NonStandard;
+            }
+
+            return NameNamespace::Domain;
+        }
+        case NameNamespace::DomainData:
+        {
+            return NameNamespace::DomainData;
+        }
+        case NameNamespace::Identity:
+        {
+            // Max id/ identifier length is 255 chars according to wiki spec.
+            // But we don't need to check for this, because that's also the max
+            // length of an identifier under the Namecoin consensus rules.
+
+            // Same as d/ regex but without IDN prefix.
+            // TODO: this doesn't exactly match the https://wiki.namecoin.org spec.
+            std::regex identityPattern("[a-z0-9]+(-[a-z0-9]+)*");
+            if (!std::regex_match (label, identityPattern))
+            {
+                return NameNamespace::NonStandard;
+            }
+
+            return NameNamespace::Identity;
+        }
+        case NameNamespace::IdentityData:
+        {
+            return NameNamespace::IdentityData;
+        }
+        case NameNamespace::NonStandard:
```

Comment:
What do you think about grouping `DomainData`, `IdentityData` and `NonStandard` as three `case` clauses with fall-through, returning `purported`?  Rather than having all three explicitly.  That seems to simplify the code and make it easier to read at least in my opinion - basically saying "these three don't have any extra conditions on the validity of the data".

Reply:
I like this.  Will do.